### PR TITLE
ADCM-2343  and ADCM-2372 API tests on /rbac/role and /rbac/policy

### PR DESCRIPTION
**Pull request by Sealwing**

### CODE REVIEW

File: tests/api/testdata/db_filler.py [hunk 1]
```diff
@@ -194,11 +197,29 @@ def _solve_field_relations(self, endpoint: Endpoints, data: dict, field: Field,
...
         else:
             raise NotImplementedError(f"Relations logic needs to be implemented for {endpoint} for field {field.name}")
 
         return _data, field
 
+    def _get_adcm_object_id_by_object_type(
+        self, object_type: Literal["cluster", "service", "component", "provider", "host"]
+    ) -> int:
+        """Get random created object by given type"""
+        return random.choice(get_endpoint_data(adcm=self.adcm, endpoint=Endpoints[object_type.capitalize()]))["id"]
```
Comment (skhomuti) [inline tests/api/testdata/db_filler.py:221]: why you do not use enum in the field `object_type`?

File: tests/api/testdata/db_filler.py [hunk 2]
```diff
@@ -248,9 +269,9 @@ def _choose_fk_field_value(self, field: Field, fk_data: list):
         """Choose a random fk value for the specified field"""
         if isinstance(field.f_type, ForeignKey):
             fk_class_name = field.f_type.fk_link.__name__
-            if fk_class_name in self._available_fkeys:
+            if available_fkeys := self._available_fkeys.get(fk_class_name, None):
```
Comment (skhomuti) [inline tests/api/testdata/db_filler.py:272]: please add a comment to explain why was it required

Comment (skhomuti) [reply tests/api/testdata/db_filler.py:272]: `if fk_vals := ...`

File: tests/api/testdata/generators.py [hunk 3]
```diff
@@ -87,6 +86,8 @@ def _fill_pytest_param(
     if endpoint.spec_link:
         marks.append(allure.link(url=endpoint.spec_link, name="Endpoint spec"))
     param_id = f"{endpoint.path.replace('/', '_')}_{method.name}_{positive_str}"
+    if sum(1 for e in Endpoints if e.path == endpoint.path) > 1:
+        param_id += f"_{endpoint.endpoint.get_data_class_pretty_name()}"
```
Comment (skhomuti) [inline tests/api/testdata/generators.py:90]: probably you can move it to the endpoint, like something `endpoint.pretty_name()`
return path by default or path + dataclass if it's needed

File: tests/api/testdata/getters.py [hunk 4]
```diff
@@ -29,11 +29,15 @@ def get_endpoint_data(adcm: ADCMTestApiWrapper, endpoint: Endpoints) -> list:
...
-        return res.json()
+    result = res.json()
     # New endpoints always return a response with pagination.
     # In the future all endpoints will return that
-    return res.json().get("results")
+    if not isinstance(result, list):
+        result = result.get("results")
+    if endpoint.endpoint.filter_predicate is None:
+        return result
+    # Is used mostly for "special" or "corner-case" endpoints
```
Comment (skhomuti) [inline tests/api/testdata/getters.py:39]: not really. it is used for endpoints with more than one dataclass
"special or corner-case" does not explain

File: tests/api/utils/types.py [hunk 5]
```diff
@@ -406,6 +419,41 @@ def __init__(self, **kwargs):
         ]
 
 
+class GenericForeignKeyList(BaseType):
```
Comment (skhomuti) [inline tests/api/utils/types.py:422]: seems like you can use `return_value` func

**Review by skhomuti — CHANGES_REQUESTED** [review]

File: tests/api/utils/types.py (hunk 5 above)
Comment (Sealwing) [reply tests/api/utils/types.py:422]: I guess it's kind of parallel thing, more of a wrapper, not a data receiver, so not sure how to use it instead of `payload` workaround

File: tests/api/testdata/getters.py (hunk 4 above)
Comment (Sealwing) [reply tests/api/testdata/getters.py:39]: Changed

File: tests/api/testdata/generators.py (hunk 3 above)
Comment (Sealwing) [reply tests/api/testdata/generators.py]: sure

File: tests/api/testdata/db_filler.py (hunk 2 above)
Comment (Sealwing) [reply tests/api/testdata/db_filler.py]: Tried to add decent explanation here

File: tests/api/testdata/db_filler.py (hunk 1 above)
Comment (Sealwing) [reply tests/api/testdata/db_filler.py:221]: not the case

File: tests/api/testdata/generators.py [hunk 6]
```diff
@@ -87,7 +87,7 @@ def _fill_pytest_param(
         marks.append(allure.link(url=endpoint.spec_link, name="Endpoint spec"))
     param_id = f"{endpoint.path.replace('/', '_')}_{method.name}_{positive_str}"
     if sum(1 for e in Endpoints if e.path == endpoint.path) > 1:
-        param_id += f"_{endpoint.endpoint.get_data_class_pretty_name()}"
+        param_id += f"_{endpoint.pretty_name}"
```
Comment (skhomuti) [inline tests/api/testdata/generators.py:90]: two lines above can be in pretty_name realization too

Comment (Sealwing) [reply tests/api/testdata/generators.py]: discussed, reworked

File: tests/api/testdata/db_filler.py [hunk 7]
```diff
@@ -194,11 +202,29 @@ def _solve_field_relations(self, endpoint: Endpoints, data: dict, field: Field,
                 ][-1]
                 field.f_type.schema = build_schema_by_json(current_config_log[field.name])
 
+        elif endpoint == Endpoints.RbacNotBuiltInPolicy:
+            if field.name == "object":
+                role_fk = _data[related_field_name]
+                role = get_object_data(adcm=self.adcm, endpoint=endpoint.RbacAnyRole, object_id=role_fk)
+                _data[field.name] = field.f_type.payload = [
```
Comment (skhomuti) [inline tests/api/testdata/db_filler.py:209]: payload is still needed?

**Review by skhomuti — CHANGES_REQUESTED** [review]

File: tests/api/testdata/db_filler.py (hunk 7 above)
Comment (Sealwing) [reply tests/api/testdata/db_filler.py:209]: Yep, I use it in other places. I guess we need to cover more than one place for such "special" stuff anyway

Comment (Sealwing) [reply tests/api/testdata/db_filler.py]: I've reworked `payload` a little, made some fixes, so it'll be called directly in `_prepare_field_value_for_object_creation`

**Review by vgorkavenko — APPROVED** [review]

**Review by skhomuti — APPROVED** [review]
